Revit_Core_Engine/Query/PanelSurfaces.cs

```diff
@@ -64,127 +118,221 @@ public static partial class Query
+                    {
+                        Dictionary<PlanarSurface, List<PlanarSurface>> subResult = new Dictionary<PlanarSurface, List<PlanarSurface>>();
+
+                        fullSolids = fullSolids.SelectMany(x => SolidUtils.SplitVolumes(x)).Where(x => x != null).ToList();
+                        if (hostObject is Wall)
                         {
-                            openings.Add(new PlanarSurface(loop.FromRevit(), null));
+                            fullSolids.ForEach(x => BooleanOperationsUtils.CutWithHalfSpaceModifyingOriginalSolid(x, objectPlanes[0]));
+                            fullSolids = fullSolids.Where(x => x != null).ToList();
+                            Autodesk.Revit.DB.Plane flippedPlane = Autodesk.Revit.DB.Plane.CreateByNormalAndOrigin(-objectPlanes[0].Normal, objectPlanes[0].Origin + objectPlanes[0].Normal * 1e-3);
+                            fullSolids.ForEach(x => BooleanOperationsUtils.CutWithHalfSpaceModifyingOriginalSolid(x, flippedPlane));
+                            fullSolids = fullSolids.Where(x => x != null).SelectMany(x => SolidUtils.SplitVolumes(x)).ToList();
+                            objectPlanes[0] = Autodesk.Revit.DB.Plane.CreateByNormalAndOrigin(-objectPlanes[0].Normal, objectPlanes[0].Origin);
                         }
 
-                        if (inserts.Count != 0)
+                        foreach (Autodesk.Revit.DB.Plane plane in objectPlanes)
                         {
-                            List<Solid> openingVolumes = new List<Solid>();
-                            foreach (Solid s2 in solidsWithOpenings)
-                            {
-                                openingVolumes.Add(BooleanOperationsUtils.ExecuteBooleanOperation(s, s2, BooleanOperationsType.Difference));
-                            }
-
-                            foreach (Solid s2 in openingVolumes)
+                            foreach (Solid s in fullSolids)
                             {
-                                foreach (Autodesk.Revit.DB.Face f in s2.Faces)
+                                List<CurveLoop> loops = new List<CurveLoop>();
+                                foreach (Autodesk.Revit.DB.Face f in s.Faces)
                                 {
                                     PlanarFace pf = f as PlanarFace;
                                     if (pf == null)
                                         continue;
 
                                     if (Math.Abs(1 - pf.FaceNormal.DotProduct(plane.Normal)) <= settings.DistanceTolerance && Math.Abs((pf.Origin - plane.Origin).DotProduct(plane.Normal)) <= settings.AngleTolerance)
+                                        loops.AddRange(pf.GetEdgesAsCurveLoops());
+                                }
+
+                                CurveLoop outline = loops.FirstOrDefault(x => x.IsCounterclockwise(plane.Normal));
+                                PlanarSurface surface = new PlanarSurface(outline.FromRevit(), null);
```

Comment:
This piece of code has been there for years already and never heard of any issues with it, so I think it should be fine. Worst case scenario, it will get caught by `catch`, which is fine taken we could not do much more with a null check (would still need to raise some sort of an error anyways).